src/openforms/appointments/contrib/base.py

```diff
@@ -0,0 +1,163 @@
+
+
+@dataclass()
+class AppointmentProduct:
+    identifier: str
+    name: str
+    code: str = None
+
+
+@dataclass()
+class AppointmentLocation:
+    identifier: str
+    name: str
+
+
+@dataclass()
+class AppointmentClient:
+    last_name: str = None
+    initials: str = None
+    birthdate: date = None
+    phonenumber: str = None
+
+
+class BasePlugin:
+    """
+    Base Appointment plugin.
+    """
+
+    verbose_name = _("Set the 'verbose_name' attribute for a human-readable name")
+    configuration_options = EmptyOptions
+
+    def __init__(self, identifier: str):
+        self.identifier = identifier
+
+    # override
+
+    def get_available_products(self, current_products: list = None) -> list:
+        """
+        Retrieve all available products and services to create an appointment
+        for. You can pass `current_products` to only retrieve available
+        products in combination with the `current_products`.
+
+        @param current_products: List of `AppointmentProduct`, as obtained from another `get_available_products` call.
+        @return: List of `AppointmentProduct`
+        """
+        raise NotImplementedError()
+
+    def get_locations(self, products: list) -> list:
+        """
+        Retrieve all available locations for given `products`.
+
+        @param products: List of `AppointmentProduct`, as obtained from `get_available_products`.
+        @return: List of `AppointmentLocation`
+        """
+        raise NotImplementedError()
+
+    def get_dates(
+        self,
+        products: list,
+        location: AppointmentLocation,
+        start_at: date = None,
+        end_at: date = None,
+    ) -> list:
+        """
+        Retrieve all available dates for given `products` and `location`.
+
+        @param products: List of `AppointmentProduct`, as obtained from `get_available_products`.
+        @param location: An `AppointmentLocation`, as obtained from `get_locations`.
+        @param start_at: The start `date` to retrieve available dates for. Default: `date.today()`.
+        @param end_at: The end `date` to retrieve available dates for. Default: 14 days after `start_date`.
+        @return: List of `date`
+        """
+        raise NotImplementedError()
+
+    def get_times(
+        self, products: list, location: AppointmentLocation, day: date
+    ) -> list:
+        """
+        Retrieve all available times for given `products`, `location` and `day`.
+
+        @param products: List of `AppointmentProduct`, as obtained from `get_available_products`.
+        @param location: An `AppointmentLocation`, as obtained from `get_locations`.
+        @param day: A `date` to retrieve available times for.
+        @return: List of `datetime`.
+        """
+        raise NotImplementedError()
+
+    def get_calendar(
+        self,
+        products: list,
+        location: AppointmentLocation,
+        start_at: date = None,
+        end_at: date = None,
+    ) -> OrderedDict:
```

Comment:
```suggestion
    def get_calendar(
        self,
        products: List[AppointmentProduct],
        location: AppointmentLocation,
        start_at: Optional[date] = None,
        end_at: Optional[date] = None,
    ) -> OrderedDict[date, List[datetime]]:
```

We can also use regular `dict` rather than `OrderedDict` since order is guaranteed to be the insertion order in modern Python versions :)